### PR TITLE
DAML-LF add Type Representation value

### PR DESCRIPTION
**Pull request by remyhaemmerle-da**
We add Type representation value to daml-lf.

This is used by the triggers instead the `to_text_type_con` primitive.

### Pull Request Checklist

- [X] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [X] Include appropriate tests
- [X] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Add a line to the [release notes](https://github.com/digital-asset/daml/blob/master/unreleased.rst), if appropriate
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: compiler/daml-lf-tools/src/DA/Daml/LF/TypeChecker/Check.hs [hunk 1]
```diff
@@ -502,24 +503,23 @@ typeOf = \case
...
+    checkSimpleType ty
+    pure $ TBuiltin BTTypeRep
   EUpdate upd -> typeOfUpdate upd
   EScenario scen -> typeOfScenario scen
   ELocation _ expr -> typeOf expr
 
 -- Check that the type contains no type variables or quantifiers
-checkAnyType :: MonadGamma m => Type -> m ()
-checkAnyType ty =
+checkSimpleType :: MonadGamma m => Type -> m ()
```
Comment (cocreature) [inline compiler/daml-lf-tools/src/DA/Daml/LF/TypeChecker/Check.hs]:
```suggestion
checkGroundType :: MonadGamma m => Type -> m ()
```

File: compiler/daml-lf-tools/src/DA/Daml/LF/TypeChecker/Check.hs [hunk 2]
```diff
@@ -502,24 +503,23 @@ typeOf = \case
...
+checkSimpleType ty =
```
Comment (cocreature) [inline compiler/daml-lf-tools/src/DA/Daml/LF/TypeChecker/Check.hs]:
```suggestion
checkGroundType ty =
```

File: compiler/daml-lf-tools/src/DA/Daml/LF/TypeChecker/Check.hs [hunk 3]
```diff
@@ -502,24 +503,23 @@ typeOf = \case
   ESome bodyType bodyExpr -> checkSome bodyType bodyExpr $> TOptional bodyType
   ENone bodyType -> checkType bodyType KStar $> TOptional bodyType
   EToAny ty bodyExpr -> do
-    checkAnyType ty
+    checkSimpleType ty
```
Comment (cocreature) [inline compiler/daml-lf-tools/src/DA/Daml/LF/TypeChecker/Check.hs]:
```suggestion
    checkGroundType ty
```

File: compiler/daml-lf-tools/src/DA/Daml/LF/TypeChecker/Check.hs [hunk 4]
```diff
@@ -502,24 +503,23 @@ typeOf = \case
...
     checkExpr bodyExpr ty
     pure $ TBuiltin BTAny
   EFromAny ty bodyExpr -> do
-    checkAnyType ty
+    checkSimpleType ty
```
Comment (cocreature) [inline compiler/daml-lf-tools/src/DA/Daml/LF/TypeChecker/Check.hs]:
```suggestion
    checkGroundType ty
```

File: daml-lf/validation/src/test/scala/com/digitalasset/daml/lf/validation/TypingSpec.scala [hunk 5]
```diff
@@ -378,7 +378,9 @@ class TypingSpec extends WordSpec with TableDrivenPropertyChecks with Matchers {
         E"Λ (τ : *). λ (t: Any) → (( to_any @(∀ (α : ⋆). Int64) t ))",
         E"Λ (τ : *). λ (t: Any) → (( to_any @(List (Optional (∀ (α : ⋆). Int64))) t ))",
         // ExpToTextTypeConName
-        E"to_text_type_con_name @Mod:NoSuchType",
+//        E"(( type_rep @Mod:NoSuchType ))",
```
Comment (cocreature) [inline daml-lf/validation/src/test/scala/com/digitalasset/daml/lf/validation/TypingSpec.scala]: What’s up with those tests?

File: daml-lf/archive/src/main/protobuf/com/digitalasset/daml_lf_dev/daml_lf_1.proto [hunk 6]
```diff
@@ -274,6 +274,10 @@ enum PrimType {
   // Builtin type 'Any'
   // *Available in versions >= 1.dev*
   ANY = 18;
```
Comment (cocreature) [inline daml-lf/archive/src/main/protobuf/com/digitalasset/daml_lf_dev/daml_lf_1.proto]: Did you forget to add the equality primitive here?

Comment (remyhaemmerle-da) [reply daml-lf/archive/src/main/protobuf/com/digitalasset/daml_lf_dev/daml_lf_1.proto]: Indeed.

Comment (remyhaemmerle-da) [reply daml-lf/archive/src/main/protobuf/com/digitalasset/daml_lf_dev/daml_lf_1.proto]: Fixing that now.

**Comment by cocreature** [comment]
Pushed an update that gets triggers to compile. To get the tests to work we need the equality primitive (currently I just call `error`).

File: compiler/daml-lf-ast/src/DA/Daml/LF/Ast/Version.hs [hunk 7]
```diff
@@ -67,11 +67,11 @@ featureAnyType = Feature
    , featureCppFlag = "DAML_ANY_TYPE"
    }
 
-featureTemplateTypeRep :: Feature
-featureTemplateTypeRep = Feature
-    { featureName = "templateTypeRep builtin"
+featureTypeRep :: Feature
```
Comment (associahedron) [inline compiler/daml-lf-ast/src/DA/Daml/LF/Ast/Version.hs]: @cocreature should `featureTypeRep` be in the `allFeatures` list below?

Comment (cocreature) [reply compiler/daml-lf-ast/src/DA/Daml/LF/Ast/Version.hs]: Good catch, fixed!

File: daml-lf/validation/src/test/scala/com/digitalasset/daml/lf/validation/TypingSpec.scala (hunk 5 above)
Comment (remyhaemmerle-da) [reply daml-lf/validation/src/test/scala/com/digitalasset/daml/lf/validation/TypingSpec.scala]: fixed

File: daml-lf/validation/src/main/scala/com/digitalasset/daml/lf/validation/Typing.scala [hunk 8]
```diff
@@ -727,35 +727,21 @@ private[validation] object Typing {
         checkExpr(exp, TScenario(typ))
     }
 
-    private def checkAnyType(typ: Type): Unit = {
-      // we check that typ contains neither variables nor quantifiers
-      TypeTraversable(typ)
+    // we check that typ contains neither variables nor quantifiers
+    private def checkSimpleType_(typ: Type): Unit = {
```
Comment (aherrmann-da) [inline daml-lf/validation/src/main/scala/com/digitalasset/daml/lf/validation/Typing.scala]: Should this also be called `checkGroundType` to be consistent with https://github.com/digital-asset/daml/pull/3326#discussion_r342036768?

File: daml-lf/validation/src/test/scala/com/digitalasset/daml/lf/validation/TypingSpec.scala [hunk 9]
```diff
@@ -166,32 +166,32 @@ class TypingSpec extends WordSpec with TableDrivenPropertyChecks with Matchers {
...
-        E"""λ (t: Any) -> (( from_any @(Mod:R Text) t ))""" ->
+        E"""λ (t: Any) → (( from_any @(Mod:R Text) t ))""" ->
           T"Any → Option (Mod:R Text)",
-        E"""λ (t: Any) -> (( from_any @Text t ))""" ->
+        E"""λ (t: Any) → (( from_any @Text t ))""" ->
           T"Any → Option Text",
-        E"""λ (t: Any) -> (( from_any @Int64 t ))""" ->
+        E"""λ (t: Any) → (( from_any @Int64 t ))""" ->
           T"Any → Option Int64",
         // ExpToTextTypeConName
```
Comment (aherrmann-da) [inline daml-lf/validation/src/test/scala/com/digitalasset/daml/lf/validation/TypingSpec.scala]:
```suggestion
        // ExpTypeRep
```

File: daml-lf/validation/src/test/scala/com/digitalasset/daml/lf/validation/TypingSpec.scala [hunk 10]
```diff
@@ -378,7 +378,9 @@ class TypingSpec extends WordSpec with TableDrivenPropertyChecks with Matchers {
         E"Λ (τ : *). λ (t: Any) → (( to_any @(∀ (α : ⋆). Int64) t ))",
         E"Λ (τ : *). λ (t: Any) → (( to_any @(List (Optional (∀ (α : ⋆). Int64))) t ))",
         // ExpToTextTypeConName
```
Comment (aherrmann-da) [inline daml-lf/validation/src/test/scala/com/digitalasset/daml/lf/validation/TypingSpec.scala]:
```suggestion
        // ExpTypeRep
```

File: daml-lf/validation/src/test/scala/com/digitalasset/daml/lf/validation/TypingSpec.scala (hunk 9 above)
Comment (remyhaemmerle-da) [reply daml-lf/validation/src/test/scala/com/digitalasset/daml/lf/validation/TypingSpec.scala]: Thx. Fixed.

File: daml-lf/validation/src/test/scala/com/digitalasset/daml/lf/validation/TypingSpec.scala (hunk 10 above)
Comment (remyhaemmerle-da) [reply daml-lf/validation/src/test/scala/com/digitalasset/daml/lf/validation/TypingSpec.scala]: fixed

File: daml-lf/validation/src/main/scala/com/digitalasset/daml/lf/validation/Typing.scala (hunk 8 above)
Comment (remyhaemmerle-da) [reply daml-lf/validation/src/main/scala/com/digitalasset/daml/lf/validation/Typing.scala]: fixed.

File: compiler/daml-lf-tools/src/DA/Daml/LF/TypeChecker/Check.hs [hunk 11]
```diff
@@ -502,24 +503,23 @@ typeOf = \case
...
     checkExpr bodyExpr (TBuiltin BTAny)
     pure $ TOptional ty
-  EToTextTypeConName tycon -> do
-    -- Ensure that the type is known.
-    _ <- inWorld (lookupDataType tycon)
-    pure $ TBuiltin BTText
+  ETypeRep ty -> do
+    checkSimpleType ty
```
[the remark this answers is not in the document]
Comment (remyhaemmerle-da) [reply compiler/daml-lf-tools/src/DA/Daml/LF/TypeChecker/Check.hs]: fixed.

**Comment by remyhaemmerle-da** [comment]
Equality builtin pushed. 
Trigger tests fail.

**Comment by cocreature** [comment]
I fixed the trigger tests. From my side this is ready to be merged.

**Pull request merged**